proposals/0000-visible-forall.rst

```diff
@@ -0,0 +1,1269 @@
+    to whether it stands for a type variable or a term variable.
+
+    Before this proposal, all term variables (retained, values, runtime) used
+    names from the term namespace, and all type variables (erased, types,
+    compile-time) used names from the type namespace. This is no longer the
+    case.
+
+T2T-Mapping
+~~~~~~~~~~~
+
+T2T (term-to-type) is a mapping from expressions to types that operates on a
+resolved syntax tree and is invoked by the ``T2T`` typing rule.
+
+The T2T mapping is partial: it succeeds on expressions that are within the
+Static Subset (introduced in `#378 Design for Dependent Types
+<https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0378-dependent-type-design.rst>`_),
+and fails on expressions outside of this subset.
+
+* Variables and constructors (regardless of their namespace) are mapped
+  directly, without modification.
+
+  * In the type checking environment, the variable must stand for a type variable,
+    or else it treated as a fresh skolem constant.
+
+  * In the type checking environment, the constructor must stand for a type
+    constructor, or else require ``DataKinds``.
+
+  * In the type checking environment, there should be no variable of the same
+    name but from a different namespace, or else raise an ambiguity error (does
+    not apply to constructors).
+
+* The types-in-terms (such as ``a -> b``, ``a => b``, ``forall a. b``) are
+  mapped to types directly, without modification aside from recursively
+  processing subterms.
+
+* Function application ``e₀ e₁`` is mapped to type-level function
+  application ``t₀ t₁``, where ``t₀ = t2t(e₀)``, ``t₁ = t2t(e₁)``.
+
+* With ``DataKinds``, a numeric literal ``42`` is mapped to a promoted numeric
+  literal.
+
+* With ``DataKinds``, a string literal ``"Hello"`` is mapped to a promoted
+  string literal ``"Hello"``.
+
+* With ``DataKinds``, a character literal ``'x'`` is mapped to a promoted
+  character literal ``'x'``.
+
+* A fractional numeric literal ``3.14`` cannot be mapped at the
+  moment, as we do not have promoted fractional numeric literals.
+
+* An unboxed numeric literal ``1337#`` cannot be mapped at the moment,
+  as we do not have promoted unboxed types.
+
+* With ``DataKinds``, a tuple ``(e₀, e₁, ...)ₑ`` is mapped to a promoted tuple
+  ``(t₀, t₁, ...)ₑ``, where ``t₀ = t2t(e₀)``, ``t₁ = t2t(e₁)``.
+
+* An unboxed tuple ``(# a, b #)`` cannot be mapped at the moment, as we do not
+  have promoted unboxed types.
+
+* With ``DataKinds``, a list literal ``[e₀, e₁, ...]`` is mapped to a promoted
+  list ``[t₀, t₁, ...]``, where ``t₀ = t2t(e₀)``, ``t₁ = t2t(e₁)``.
+
+* With ``TypeApplications``, type application ``e₀ @t₁`` is mapped to
+  type-level type application ``t₀ @t₁``, where ``t₀ = t2t(e₀)``.
+
+* With ``TypeOperators``, infix application ``e₀ op e₁`` is mapped to
+  type-level infix application ``e₀ tyop e₁``, where ``t₀ = t2t(e₀)``, ``t₁ =
+  t2t(e₁)``, ``tyop = t2t(op)``.
```

Comment:
I don't think we need `tyop = t2t(op)`. Operators are always plain names, which are unaffected by the translation. This point might want to note that no re-association is done, in case a type-level `+` has different fixity than the term-level `+`. (This non-reassociation is pretty clearly implied by everything else here, but it's still nice to remind the reader.)